### ISSUE TITLE
Best practice for waiting until a simulator is ready for usage

### ISSUE THREAD

**Issue by talkol** [facebook/FBSimulatorControl#295]
Our currently naive approach to run an e2e test starts with:

`fbsimctl --state=shutdown --state=shutting-down "iPhone 6" "iOS 9.3" boot`

This usually works, but we're experiencing some flakiness where the simulator isn't always ready to run the tests after the command returns.

One example that reproduces an error is this: if you reset the simulator (`Reset Content and Settings...`) and close it, the next launch takes a while as the simulator installation bar is shown. The `boot` command returns but the simulator isn't ready.

Another example is if the simulator is in the `creating` state. When does this state end?

TLDR What's the recommended approach to wait for a working simulator?

**Comment by lawrencelomax** [comment]
The information that we get is directly from `CoreSimulator.framework`. We can augment `boot` with additional heuristics for 'readyness' of the Simulator but we would first need to establish what they are. What kind of flakyness are you seeing, or what commands are you trying to run against the half-booted Simulator? It could be that your command is dependent on some daemon of the `launchd_sim` that isn't currently in a stable state. `FBSimulatorControl` [already has some heuristics](https://github.com/facebook/FBSimulatorControl/blob/master/FBSimulatorControl/Strategies/FBSimulatorBootStrategy.m#L345), but these may be inadequate for your case.

I'd also suggest using UDIDs, in the event that there are multiple Simulators matching that query. You can get this output from `fbsimctl list`, select one and pass it into the next invocation of `fbsimctl UDID boot`. You will end up waiting on a Simulator that is 'shutting down', to be shut down before booting it again.

**Comment by lawrencelomax** [comment]
> Another example is if the simulator is in the creating state. When does this state end?

This will then transition to `shutdown`. Erasing a Simulator will put it into the `creating` state, then to `shutdown`. Erasing is very similar to creating a new Simulator in this way.

> TLDR What's the recommended approach to wait for a working simulator?

`boot` will wait for `CoreSimulator` then perform some additional checks.

**Comment by talkol** [comment]
Hi, I took your recommendation of using UDID and this approach indeed reduces problems since we have less collisions between concurrent builds (each build agent has its own set of UDID's).

Also, it seems that the core reason for most of the flakiness was that one of the build machines is insanely slow (not a very old Mac Mini, but simulators running on it take x10 time to start up). Once we removed it, I see that the situation is much less problematic than originally thought.

In summary, let's leave it close and reopen if we have something intelligent to add
